### PR TITLE
Improve the `SyntaxError` constructor; add `SyntaxError.print_file_and_line`

### PR DESCRIPTION
**Pull request by tungol**
I set out to just do `print_file_and_line`, but figured out how the constructor worked along the way.

### CODE REVIEW

File: stdlib/builtins.pyi [hunk 1]
```diff
@@ -1963,14 +1963,33 @@ class StopAsyncIteration(Exception):
 
 class SyntaxError(Exception):
     msg: str
+    filename: str | None
     lineno: int | None
     offset: int | None
     text: str | None
-    filename: str | None
+    # Runtime displays errors differently if this attribute exists on the exception.
```
Comment (AlexWaygood) [inline stdlib/builtins.pyi:1970]:
```suggestion
    # Errors are displayed differently if this attribute exists on the exception.
```

File: stdlib/builtins.pyi [hunk 2]
```diff
@@ -1963,14 +1963,33 @@ class StopAsyncIteration(Exception):
...
+    # The value is always None.
```
Comment (AlexWaygood) [inline stdlib/builtins.pyi:1971]: interesting... is this documented anywhere?

**Comment by tungol** [comment]
Draft until I have a chance to look at those  mypy primer hits

**Review by AlexWaygood — COMMENTED** [review]
Nice, great investigation! What a weird constructor at runtime

**Comment by tungol** [comment]
mypy primer:

Black has a filename that's typed as `Union[str, PathLike[str]]`. I think this hit is legitimate:

```python
from traceback import print_exception

class foo:
    def __fspath__(self) -> str:
        return "filename"

se = SyntaxError("hi", (foo(), 1, 2, "this"))
print_exception(se)
```
gives this output:
```
  File "<__main__.foo object at 0x102836ed0>", line 1
    this
     ^
SyntaxError: hi
```
Which isn't very good.

setuptools is taking one exception and passing it as the first argument to SyntaxError.

```python
from traceback import print_exception

ve = ValueError("bar")
se = SyntaxError(ve)
print_exception(se)
```
gives this output:
```
SyntaxError: bar
```

Which seems probably reasonable to support?

**Comment by tungol** [comment]
Compromise: If you're not passing in an info tuple: accept whatever as the first argument. If you are passing an info tuple, expect a string. My thinking is that single-argument, like no-argument, probably doesn't care too much about the niceties.

**Review by AlexWaygood — APPROVED** [review]
Thanks!

**Pull request merged**